lib/brexit_checker/results_audiences.rb

```diff
@@ -0,0 +1,56 @@
+class BrexitChecker::ResultsAudiences
+  class << self
+    def populate_business_groups(audience_actions, selected_criteria)
+      return [] if audience_actions.blank? || selected_criteria.blank?
+
+      {
+        actions: audience_actions,
+        criteria: filter_criteria_by_actions(audience_actions, selected_criteria),
+      }
+    end
+
+    def populate_citizen_groups(audience_actions, selected_criteria)
+      return [] if audience_actions.blank? || selected_criteria.blank?
+
+      all_possible_groups = BrexitChecker::Groups.load_all
+      grouped_actions = all_possible_groups.each_with_index.map do |group, group_index|
+        actions_in_group = filter_actions_by_group(audience_actions, group.key)
+        if actions_in_group.empty?
+          nil
+        else
+          {
+            group: group,
+            group_index: group_index + 1,
+            actions: actions_in_group,
+            criteria: filter_criteria_by_actions(actions_in_group, selected_criteria),
+          }
+        end
+      end
+      grouped_actions.compact
+    end
+
+  private
+
+    def filter_actions_by_group(actions, group_key)
+      actions.select { |action| action.grouping_criteria.include?(group_key) }
+    end
+
+    def extract_groupings(actions)
+      actions.flat_map(&:grouping_criteria).uniq
+    end
+
+    def filter_actions_by_audience(audience, actions)
+      actions.group_by(audience)
+    end
+
+    def filter_criteria_by_actions(actions, criteria)
+      return [] if actions.empty? || criteria.empty?
+
+      action_criteria = actions.flat_map do |action|
+        BrexitChecker::Criteria::Extractor.extract(action.criteria).to_a
```

Comment:
the extractor was previously only used for unit tests I believe. Maybe it would make sense to move that functionality to the Action model?

Reply:
can we check in on this one?

Reply:
perhaps same as above for this one.

This is really about filtering some pre-selected actions by page criteria.
I think before we get to this code, we've already taken the page criteria and filtered down to just the relevant actions on this page. What we need to do now is take actions that have already been grouped and find out which criteria are relevant to that group